src/VisualStudio/Core/Def/Implementation/FindUsages/VisualStudioFindSymbolMonikerUsagesService.cs

```diff
@@ -0,0 +1,121 @@
+
+                // If we got no items, we're done.
+                if (referenceItems.Length == 0)
+                    break;
+
+                foreach (var item in referenceItems)
+                    yield return item;
+
+                // Otherwise, we got some items.  Return them to our caller and attempt to retrieve
+                // another page.
+                currentPage++;
+            }
+        }
+
+        private async Task<ImmutableArray<ExternalReferenceItem>> FindReferencesByMonikerAsync(
+            ICodeIndexProvider codeIndexProvider, DefinitionItem definition, ImmutableArray<ISymbolMoniker> monikers,
+            IStreamingProgressTracker progress, int pageIndex, CancellationToken cancellationToken)
+        {
+            // Let the find-refs window know we have outstanding work
+            await using var _1 = await progress.AddSingleItemAsync().ConfigureAwait(false);
+
+            var results = await codeIndexProvider.FindReferencesByMonikerAsync(
+                monikers, includeDeclaration: false, pageIndex: pageIndex, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            using var _2 = ArrayBuilder<ExternalReferenceItem>.GetInstance(out var referenceItems);
+
+            foreach (var result in results)
+                referenceItems.Add(ConvertResult(definition, result));
+
+            return referenceItems.ToImmutable();
+        }
+
+        private ExternalReferenceItem ConvertResult(DefinitionItem definition, JObject obj)
+        {
+            var projectName = obj.Value<string>("projectName");
+            var displayPath = obj.Value<string>("displayPath");
+            var span = ConvertLinePositionSpan(obj.Value<JObject>("range"));
+            var text = obj.Value<string>("text");
+
+            return new CodeIndexExternalReferenceItem(
+                this, definition, obj, projectName, displayPath, span, text);
+
+            static LinePositionSpan ConvertLinePositionSpan(JObject obj)
+                => new LinePositionSpan(
+                    ConvertLinePosition(obj.Value<JObject>("start")),
+                    ConvertLinePosition(obj.Value<JObject>("end")));
+
+            static LinePosition ConvertLinePosition(JObject obj)
+                => new LinePosition(obj.Value<int>("line"), obj.Value<int>("character"));
+        }
+
+        private ImmutableArray<ISymbolMoniker> ConvertMonikers(ImmutableArray<SymbolMoniker> monikers)
+            => monikers.SelectAsArray(ConvertMoniker);
+
+        private ISymbolMoniker ConvertMoniker(SymbolMoniker moniker)
+            => new MonikerWrapper(moniker);
+
+        private CancellationToken CancelLastNavigationAndGetNavigationToken()
+        {
+            lock (_gate)
+            {
+                _lastNavigationCancellationSource?.Cancel();
+                _lastNavigationCancellationSource = new CancellationTokenSource();
+                return _lastNavigationCancellationSource.Token;
```

Comment:
I think @sharwell wrote a helper for this pattern since we do it so much.

Reply:
couldn't find anything.  talked to sam.  there are other libs, but nothing in VS.